### DISCUSSION TITLE
sending a http request without freezing the console

### DISCUSSION THREAD

**Discussion by lukas-runge** [hossimo/GMA3Plugins#25] [Q&A]
Hey everyone,

as already stated in my welcome post I am working on porting the [MA2 REAPER Live Marker Sync](https://github.com/locimation/ma2-tc-import/tree/master/TC%20Live)-Plugin from @locimation to MA3: [liveConnect2Reaper](https://github.com/gegenLichtLindenfels/liveConnect2Reaper)

I am currently facing an issue where every **http request takes at least 2 seconds** in which the MA3 onPC is **completely frozen** (no user input is processed). Is there a way to get the http request time down to less that 1/10 of a second or alternatively **process it in the background without freezing the console**?

Happy to hear if anyone has a possible solution already!
Kind regards
@lukas-runge

**Comment by hossimo** [comment]
You probably need to process the HTTP request async or yield until something comes back. My [USB Stomper](https://github.com/hossimo/GMA3Plugins/blob/088b89fdc3e90be532d81f62daa9153baceb8728/grandMA3/shared/resource/lib_plugins/USB%20Stomper.lua) and [Fade Master](https://github.com/hossimo/GMA3Plugins/blob/088b89fdc3e90be532d81f62daa9153baceb8728/grandMA3/shared/resource/lib_plugins/Fade%20Master.lua) have examples of this.

That way you can send the request and yield every loop until you get the response.

**Reply by lukas-runge** [reply]
I'm already using coroutine.yield(time) to allow some processing time between executes. Here is a minimal example:

```lua
-- load HTTP package
local http = require("http")

function Main()
  while true do
    -- get marker response from reaper
    local response = http.request('http://localhost:18080/_/MARKER')  
    
    -- print response
    printf(response)

    -- allow other processes for 3 seconds
    coroutine.yield(3)
  end
end

return Main
```

It seems like the actual duration of the `http.request` triggers the freezing. Can I take care of that somehow by running it in parallel? Do you know how that would work?

**Thank you for your support!** @hossimo 
Kind regards
@lukas-runge

**Reply by hossimo** [reply]
I see, I haven't looked into the included http class but it seems request is blocking. Perhaps something like [lua-http](https://daurnimator.github.io/lua-http/0.3/) would be a better fit?

**Reply by lukas-runge** [reply]
Great idea, this library looks like its a perfect fit for me - if i'd only knew how to install such library inside the MA. 😬 Their GitHub says it should be installed via `luarocks`. Is there a way to use / install `luarocks` on the MA? I am a total newb to lua librarys and how to integrate them into the LUA environment inside the MA. 😅

**Reply by hossimo** [reply]
Probably not easily. Although looking at `C:\ProgramData\MALightingTechnology\gma3_1.4.0\shared\resource\lib_plugins\requirements\http.lua` it looks like they are using the library from [luasocket](https://github.com/diegonehab/luasocket). I'm not up on the module but it looks like `socket.http.TIMEOUT` is a property that is set to 60 seconds by default, Not sure if you could lower the timeout to something more manageable. There seems to be talk about a non blocking mode, but I did not see anything in the documentation, though I didn't look too hard. Might be worth an ask to them though.
